cc/mcts_player.cc

```diff
@@ -0,0 +1,247 @@
+    return c;
+  }
+
+  // Select from the first kN * kN moves (instead of kNumMoves) to avoid
+  // randomly choosing to pass early on in the game.
+  std::array<float, kN * kN> cdf;
+
+  cdf[0] = root_->child_N(0);
+  for (size_t i = 1; i < cdf.size(); ++i) {
+    cdf[i] = cdf[i - 1] + root_->child_N(i);
+  }
+  float norm = 1 / cdf[cdf.size() - 1];
+  for (size_t i = 0; i < cdf.size(); ++i) {
+    cdf[i] *= norm;
+  }
+  float e = rnd_();
+  Coord c = SearchSorted(cdf, e);
+  std::cout << "Picked rnd(" << e << ") " << c << "\n";
+  assert(root_->child_N(c) != 0);
+  return c;
+}
+
+void MctsPlayer::TreeSearch(int batch_size) {
+  int max_iterations = batch_size * 2;
+
+  // TODO(tommadams): Avoid creating this vector each time.
+  std::vector<MctsNode*> leaves;
+  for (int i = 0; i < max_iterations; ++i) {
+    auto* leaf = root_->SelectLeaf();
+    if (leaf == nullptr) {
+      continue;
+    }
+    if (leaf->position.is_game_over() ||
+        leaf->position.n() >= kMaxSearchDepth) {
+      float value = leaf->position.CalculateScore() > 0 ? 1 : -1;
+      leaf->IncorporateEndGameResult(value, root_);
+    } else {
+      leaf->AddVirtualLoss(root_);
+      leaves.push_back(leaf);
+      if (static_cast<int>(leaves.size()) == batch_size) {
+        break;
+      }
+    }
+  }
+
+  if (!leaves.empty()) {
+    // TODO(tommadams): Avoid creating these vectors each time.
+    std::vector<const DualNet::BoardFeatures*> features;
+    features.reserve(leaves.size());
+    for (auto* leaf : leaves) {
+      features.push_back(&leaf->features);
+    }
+
+    std::vector<DualNet::Output> outputs;
+    outputs.resize(leaves.size());
+
+    RunMany(features, {outputs.data(), outputs.size()});
+
+    for (size_t i = 0; i < leaves.size(); ++i) {
+      MctsNode* leaf = leaves[i];
+      const auto& output = outputs[i];
+      leaf->RevertVirtualLoss(root_);
+      leaf->IncorporateResults(output.policy, output.value, root_);
+    }
+  }
+}
+
+bool MctsPlayer::ShouldResign() const {
+  return root_->Q_perspective() < options_.resign_threshold;
+}
+
+void MctsPlayer::PlayMove(Coord c) {
+  PushHistory();
+
+  root_ = root_->MaybeAddChild(c);
+  // Don't need to keep the parent's children around anymore because we'll
+  // never revisit them.
+  root_->parent->PruneChildren(c);
+}
+
+void MctsPlayer::PushHistory() {
+  history_.emplace_back();
+  History& history = history_.back();
+  history.comment = root_->Describe();
+  history.Q = root_->Q();
+
+  // Convert child visit counts to a probability distribution, pi.
+  // For moves before the temperature cutoff, exponentiate the probabilities by
+  // a temperature slightly larger than unity to encourage diversity in early
+  // play and hopefully to move away from 3-3s.
+  if (root_->position.n() < temperature_cutoff_) {
+    // Squash counts before normalizing.
+    for (int i = 0; i < kNumMoves; ++i) {
+      history.search_pi[i] = std::pow(root_->child_N(i), 0.95);
```

Comment:
Squash is now "0.98," which we arrived at by a combination of superstition and arcane ritualistic incantations